File: tests/c/test_sparse_writes.cpp
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: LicenseRef-NvidiaProprietary
//
// Public ovstage sparse-write test: index_map (a gather over a subset of the
// query's prims) and mask (a bitmask over the target rows) each update only their
// targeted prims, leaving the rest unchanged. CPU-only. The write-flavors example
// is the workflow tour; this file asserts the sparse semantics.

#include <ovstage/ovstage.h>
#include <ovx/path_dictionary/path_dictionary.h>
#include <ovx/path_dictionary/path_dictionary_utils.h>
#include <dlpack/dlpack.h>

#include <cstdint>
#include <cstring>

#include <gtest/gtest.h>

namespace
{

ovx_string_t str(const char* s)
{
    return ovx_string_t{ s, std::strlen(s) };
}

bool waitOp(ovstage_instance_t* stage, ovstage_enqueue_result_t enq, const char* what)
{
    if (enq.status != OVSTAGE_OK)
    {
        ADD_FAILURE() << what << " enqueue rejected (code " << enq.status << ")";
        return false;
    }
    ovstage_op_wait_result_t wait{};
    const ovstage_api_status_t werr = ovstage_wait_op(stage, enq.op_index, OVSTAGE_TIMEOUT_INFINITE, &wait);
    for (size_t i = 0; i < wait.error_op_id_count; ++i)
    {
        const ovx_string_t e = ovstage_get_last_op_error(stage, wait.error_op_ids[i]);
        ADD_FAILURE() << what << " op failed: " << std::string(e.ptr ? e.ptr : "", e.ptr ? e.length : 0);
    }
    ovstage_release_op(stage, enq.op_index);
    if (werr != OVSTAGE_OK)
        ADD_FAILURE() << what << " wait failed (code " << werr << ")";
    return werr == OVSTAGE_OK && wait.error_op_id_count == 0;
}

class SparseWriteTest : public ::testing::Test
{
protected:
    void SetUp() override
    {
        ovstage_instance_desc_t desc{};
        desc.name = "test.ovstage.sparse";
        ASSERT_EQ(ovstage_create_instance(&desc, &stage_), OVSTAGE_OK);
        dict_ = ovstage_get_path_dictionary(stage_);
        ASSERT_NE(dict_, nullptr);

        const ovx_string_t attrName = str("sparse-heat");
        ASSERT_EQ(path_dictionary_create_tokens_from_strings(dict_, &attrName, 1, &heat_).status, OVX_API_SUCCESS);
        const ovx_string_t paths[] = { str("/World/Sparse/A"), str("/World/Sparse/B"), str("/World/Sparse/C"),
                                       str("/World/Sparse/D") };
        ASSERT_EQ(path_dictionary_create_path_list_from_strings(dict_, paths, 4, &pathList_).status, OVX_API_SUCCESS);
        ASSERT_EQ(ovstage_query_from_path_list(stage_, pathList_, &query_), OVSTAGE_OK);
    }

    void TearDown() override
    {
        if (stage_ && query_ != OVSTAGE_INVALID_QUERY_HANDLE)
            waitOp(stage_, ovstage_release_query(stage_, query_), "release_query");
        if (dict_ && pathList_ != OVX_INVALID_PRIMPATH_LIST)
            path_dictionary_release_path_list_reference(dict_, pathList_);
        if (stage_)
            ovstage_destroy_instance(stage_);
    }

    void seal(ovstage_ordinal_t ordinal)
    {
        ovstage_write_floor_desc_t floor{};
        floor.ordinal = ordinal;
        floor.scope = OVSTAGE_SCOPE_ALL;
        ASSERT_TRUE(waitOp(stage_, ovstage_advance_write_floor(stage_, &floor), "advance_write_floor"));
    }

    // Latest value per prim into out[list-index]. A latest read of a sparsely
    // updated column may arrive as several groups (rows last written at
    // different ordinals); each group's prim index is a list-relative index into
    // the query's path list (our A/B/C/D order).
    void readFour(ovstage_ordinal_t endOrdinal, float out[4])
    {
        for (int i = 0; i < 4; ++i)
            out[i] = -1.0f;
        ovstage_ordinal_range_t range{};
        range.end_ordinal = endOrdinal;
        ovstage_read_handle_t read = OVSTAGE_INVALID_READ_HANDLE;
        ASSERT_TRUE(waitOp(stage_, ovstage_read_attributes(stage_, query_, &heat_, 1, range, &read), "read"));
        for (;;)
        {
            ovstage_read_group_t group{};
            const ovstage_api_status_t fetched = ovstage_fetch_read_next(stage_, read, OVSTAGE_TIMEOUT_INFINITE, &group);
            if (fetched == OVSTAGE_ERROR_END_OF_ITERATION)
                break;
            if (fetched != OVSTAGE_OK)
            {
                ADD_FAILURE() << "fetch_read_next failed (code " << fetched << ")";
                break;
            }
            if (!group.is_delete && group.data.tensor_count == 1 && group.data.tensors[0].data)
            {
                const float* rows = static_cast<const float*>(group.data.tensors[0].data);
                for (uint32_t local = 0; local < group.prims.count; ++local)
                {
                    const size_t listIndex = group.prims.index_map ? group.prims.index_map[local]
                                                                   : group.prims.offset + local;
                    const uint32_t row = group.data.index_map ? group.data.index_map[local] : local;
                    if (listIndex < 4)
                        out[listIndex] = rows[row];
                }
            }
            ovstage_release_group(stage_, &group);
        }
        waitOp(stage_, ovstage_release_read(stage_, read), "release_read");
    }

    ovstage_instance_t* stage_ = nullptr;
    path_dictionary_instance_t* dict_ = nullptr;
    ovx_token_t heat_ = OVX_INVALID_TOKEN;
    ovx_primpath_list_t pathList_ = OVX_INVALID_PRIMPATH_LIST;
    ovstage_query_handle_t query_ = OVSTAGE_INVALID_QUERY_HANDLE;
};

TEST_F(SparseWriteTest, IndexMapAndMask)
{
    int64_t four[] = { 4 };
    int64_t three[] = { 3 };
    int64_t strides[] = { 1 };
    auto tensor = [&](float* data, int64_t* shape) {
        DLTensor t{};
        t.data = data;
        t.device = { kDLCPU, 0 };
        t.ndim = 1;
        t.dtype = { kDLFloat, 32, 1 };
        t.shape = shape;
        t.strides = strides;
        return t;
    };

    float dense[] = { 10.0f, 20.0f, 30.0f, 40.0f };
    DLTensor denseTensor = tensor(dense, four);
    ovstage_write_data_t denseWrite{};
    denseWrite.tensors = &denseTensor;
    denseWrite.tensor_count = 1;
    denseWrite.is_array = false;
    ASSERT_TRUE(waitOp(stage_,
        ovstage_write_attribute(stage_, query_, { heat_, {} }, 1, denseWrite, OVSTAGE_PRIM_MODE_UPSERT), "dense"));
    seal(1);
    float values[4] = {};
    readFour(1, values);
    EXPECT_FLOAT_EQ(values[0], 10.0f);
    EXPECT_FLOAT_EQ(values[3], 40.0f);

    // [snippet:sparse-index-map-and-mask-c]
    // index_map is a GATHER: count selects how many of the query's prims are
    // targeted (the first `count`, in query order) and index_map[i] picks the
    // SOURCE tensor row for target i. A non-involutory 3-cycle proves the direction:
    // gather reads source[index_map[i]] into target i (scatter would give a different
    // result). mask is a bitmask over the `count` target rows: only set bits are
    // written. Untouched prims keep their earlier value.
    float gathered[] = { 111.0f, 222.0f, 333.0f };
    const uint32_t indexMap[] = { 1, 2, 0 }; // A<-row1(222), B<-row2(333), C<-row0(111)
    DLTensor gatheredTensor = tensor(gathered, three);
    ovstage_write_data_t gatherWrite{};
    gatherWrite.tensors = &gatheredTensor;
    gatherWrite.tensor_count = 1;
    gatherWrite.is_array = false;
    gatherWrite.count = 3;
    gatherWrite.index_map = indexMap;
    ASSERT_TRUE(waitOp(stage_,
        ovstage_write_attribute(stage_, query_, { heat_, {} }, 2, gatherWrite, OVSTAGE_PRIM_MODE_UPSERT), "gather"));
    seal(2);

    readFour(2, values);
    EXPECT_FLOAT_EQ(values[0], 222.0f);
    EXPECT_FLOAT_EQ(values[1], 333.0f);
    EXPECT_FLOAT_EQ(values[2], 111.0f);
    EXPECT_FLOAT_EQ(values[3], 40.0f);

    float masked[] = { 0.0f, 0.0f, 0.0f, 44.0f };
    const uint64_t maskBits[] = { 0b1000ull }; // bit 3 set -> only D (row 3)
    DLTensor maskedTensor = tensor(masked, four);
    ovstage_write_data_t maskWrite{};
    maskWrite.tensors = &maskedTensor;
    maskWrite.tensor_count = 1;
    maskWrite.is_array = false;
    maskWrite.count = 4;
    maskWrite.mask = maskBits;
    ASSERT_TRUE(waitOp(stage_,
        ovstage_write_attribute(stage_, query_, { heat_, {} }, 3, maskWrite, OVSTAGE_PRIM_MODE_UPSERT), "mask"));
    seal(3);
    // [/snippet:sparse-index-map-and-mask-c]

    readFour(3, values);
    EXPECT_FLOAT_EQ(values[0], 222.0f);
    EXPECT_FLOAT_EQ(values[1], 333.0f);
    EXPECT_FLOAT_EQ(values[2], 111.0f);
    EXPECT_FLOAT_EQ(values[3], 44.0f);
}

} // namespace

File: tests/python/test_sparse_writes.py
# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: LicenseRef-NvidiaProprietary
#
# Public ovstage sparse-write test: index_map (a gather over a subset of the
# query's prims) and mask (a bitmask over the target rows) each update only their
# targeted prims, leaving the rest unchanged. CPU-only. The write-flavors example
# is the workflow tour; this file asserts the sparse semantics.

import numpy as np
import pytest

from ovstage import ErrorCode, OrdinalRange, PathDictionary


def _values(stage, query, attr, end_ordinal, order):
    """Latest committed value per prim of a 1-lane column, indexed like `order`."""
    read = stage.read_attributes(query, [attr], OrdinalRange.latest(end_ordinal))
    read.wait()
    rows = {}
    for group in read.groups():
        if not group.is_delete:
            values = group.array(0)
            for local in range(group.prim_count):
                row = group.data_row_index(local) if group.has_data_index_map else local
                rows[group.prim_index(local)] = float(values[row])
        stage.release_group(group)
    read.release().wait()
    return [rows.get(i) for i in range(len(order))]


def test_index_map_and_mask(stage):
    with PathDictionary(stage) as paths:
        order = ["/World/Sparse/A", "/World/Sparse/B", "/World/Sparse/C", "/World/Sparse/D"]
        plist = paths.create_path_list_from_strings(order)
        query = stage.query_from_path_list(plist)
        try:
            heat = paths.intern_token("sparse-heat")
            stage.write_attribute(
                query, heat, ordinal=1, tensors=np.array([10.0, 20.0, 30.0, 40.0], np.float32),
                is_array=False,
            ).wait()
            stage.advance_write_floor(ordinal=1).wait()
            assert _values(stage, query, heat, 1, order) == [10.0, 20.0, 30.0, 40.0]

            # [snippet:sparse-index-map-and-mask]
            # Two sparse forms, mutually exclusive per write. index_map is a GATHER:
            # count selects how many of the query's prims are targeted (the first
            # `count`, in query order) and index_map[i] picks the SOURCE tensor row
            # for target i. A non-involutory 3-cycle proves the direction (gather reads
            # source[index_map[i]] into target i; scatter would give a different result).
            # mask is a bitmask over the `count` target rows: only set bits are written.
            # Untouched prims keep their earlier value.
            stage.write_attribute(
                query, heat, ordinal=2, tensors=np.array([111.0, 222.0, 333.0], np.float32),
                is_array=False, index_map=[1, 2, 0],  # A<-row1(222), B<-row2(333), C<-row0(111)
            ).wait()
            stage.advance_write_floor(ordinal=2).wait()
            assert _values(stage, query, heat, 2, order) == [222.0, 333.0, 111.0, 40.0]

            stage.write_attribute(
                query, heat, ordinal=3, tensors=np.array([0.0, 0.0, 0.0, 44.0], np.float32),
                is_array=False, mask=[0b1000], count=4,  # bit 3 set -> only D (row 3)
            ).wait()
            stage.advance_write_floor(ordinal=3).wait()
            # [/snippet:sparse-index-map-and-mask]

            assert _values(stage, query, heat, 3, order) == [222.0, 333.0, 111.0, 44.0]
        finally:
            stage.release_query(query).wait()
            paths.destroy_path_list(plist)


def test_index_map_entries_must_be_in_range(stage):
    """An index_map entry past the transported rows is a rejected request.

    index_map gathers SOURCE rows, so an entry is only meaningful if the payload
    actually transports that row. Out-of-range entries are rejected as invalid
    arguments rather than silently redefining how the payload is cut into rows.
    """
    with PathDictionary(stage) as paths:
        order = ["/World/Range/A", "/World/Range/B"]
        plist = paths.create_path_list_from_strings(order)
        query = stage.query_from_path_list(plist)
        try:
            heat = paths.intern_token("range-heat")

            # One transported row, two prims: broadcasting it is in range.
            stage.write_attribute(
                query, heat, ordinal=1, tensors=np.array([10.0], np.float32),
                is_array=False, index_map=[0, 0], count=2,
            ).wait()
            stage.advance_write_floor(ordinal=1).wait()
            assert _values(stage, query, heat, 1, order) == [10.0, 10.0]

            # Row 1 does not exist in a single-row payload.
            rejected = stage.write_attribute(
                query, heat, ordinal=2, tensors=np.array([20.0], np.float32),
                is_array=False, index_map=[1], count=1,
            )
            assert rejected.status == ErrorCode.INVALID_ARGUMENT
            assert "index_map" in rejected.error_message()

            # mask is how a caller targets a subset of the query's prims, and it
            # requires an explicit count.
            with pytest.raises(ValueError):
                stage.write_attribute(
                    query, heat, ordinal=2, tensors=np.array([10.0, 20.0], np.float32),
                    is_array=False, mask=[0b10],
                )
        finally:
            stage.release_query(query).wait()
            paths.destroy_path_list(plist)
